### PR TITLE
Modification of section data, creating new symbols, other improvements

### PR DESCRIPTION
**Pull request by zygzagZ**
Hello
I've had a college project where I have benefitted from your library, though I have quickly realised it was not going to be enough.
These changes are just as I have sent the project in yesterday and they still lack a bit of refinement, I am going to refactor it.

I ran the tests and they passed, but the functionality implemented in this PR was only tested on ET_REL linker object files and lacks tests.

Summary:
- allow preloading of data (build all structures in memory)
- allow modification and rebuilding ELF files – adding, deleting, reordering sections, symbols, relocations
- fixed a bug where elf header's `e_ident`'s attributes cannot be written, as they cannot be packed into int by the bindata patches-gathering monkeypatch

Implementation decisions:
- no idea what is LazyArray for, will try to get rid of it in refactor and still achieve lazy loading (even though the whole elf is loaded into memory when saving)
- created new Enums that are a bit better, but also bad. Needs some thinking
- properly splitting the functionality of appending symbols requires classes to have access to relative parent (eg. `Relocation` to a `RelocationSection` and `RelocationSection` to an `ELFFile`), things get complicated because this causes loops that break `loaded_headers`. Worked around it with lambdas containing variables, but might consider appending loaded headers to some elffile internal set structure (thus fixing "bad idea..." loaded headers behavior)

Please do not hesitate to comment in case you spot anything i forgot to mention and that would need fixing in the code.

### CODE REVIEW

**Comment by zygzagZ** [comment]
Code cleaned up, rubocop passes and so do tests. Fails on coverage because of the tool crashing (?), no idea how to fix that

Introduced methods still lack comments and documentation.

File: lib/elftools/constants.rb [hunk 1]
```diff
@@ -250,6 +250,27 @@ module SHN
...
+      SHF_MERGE = (1 << 4) # Might be merged
+      SHF_STRINGS = (1 << 5) # Contains nul-terminated strings
+      SHF_INFO_LINK = (1 << 6) # `sh_info' contains SHT index
+      SHF_LINK_ORDER = (1 << 7) # Preserve order after combining
+      SHF_OS_NONCONFORMING = (1 << 8) # Non-standard OS specific handling required
+      SHF_GROUP = (1 << 9) # Section is member of a group.
+      SHF_TLS = (1 << 10) # Section hold thread-local data.
+      SHF_COMPRESSED = (1 << 11) # Section with compressed data.
+      SHF_MASKOS = 0x0ff00000	# OS-specific.
+      SHF_MASKPROC = 0xf0000000	# Processor-specific
```
Comment (david942j) [inline lib/elftools/constants.rb:267]: replace with one space `0xf0000000 #`

File: lib/elftools/constants.rb [hunk 2]
```diff
@@ -250,6 +250,27 @@ module SHN
     end
     include SHN
 
+    # Section header types, records in +sh_type+.
```
Comment (david942j) [inline lib/elftools/constants.rb:253]: replace with proper description

File: lib/elftools/elf_file.rb [hunk 3]
```diff
@@ -143,6 +143,7 @@ def sections
     #   If +n+ is out of bound, +nil+ is returned.
     def section_at(n)
       @sections ||= LazyArray.new(num_sections, &method(:create_section))
+      @sections[n].index = n if @sections[n]
       @sections[n]
```
Comment (david942j) [inline lib/elftools/elf_file.rb:147]: suggest
`@sections[n].tap { |sec| sec.index = n if sec }`

File: lib/elftools/elf_file.rb [hunk 4]
```diff
@@ -321,6 +333,54 @@ def save(filename)
       IO.binwrite(filename, all)
     end
 
+    def rebuild
```
Comment (david942j) [inline lib/elftools/elf_file.rb:336]: add description on what does this function do and what it returns

File: lib/elftools/elf_file.rb [hunk 5]
```diff
@@ -321,6 +333,54 @@ def save(filename)
...
+      all = header.to_binary_s
+
+      sections.each do |s|
+        s.rebuild
+
+        next if s.size <= 0
+
+        if s.header.sh_addralign != 0
+          # align with 0 bytes
```
Comment (david942j) [inline lib/elftools/elf_file.rb:345]: null bytes

File: lib/elftools/elf_file.rb [hunk 6]
```diff
@@ -321,6 +333,54 @@ def save(filename)
...
+
+        s.header.sh_offset = all.size
+        all += s.data
+      end
+
+      header.e_shnum = sections.size
+      header.e_shoff = all.size
+
+      sections.each do |s|
+        sh = s.header.to_binary_s
```
Comment (david942j) [inline lib/elftools/elf_file.rb:357]: variable `sh` can be omitted

File: lib/elftools/elf_file.rb [hunk 7]
```diff
@@ -321,6 +333,54 @@ def save(filename)
...
+        all += sh
+      end
+
+      all[0...header.num_bytes] = header.to_binary_s
```
Comment (david942j) [inline lib/elftools/elf_file.rb:361]: add a comment about why this line is needed (something like "fields of `header` were changed ... ")

File: lib/elftools/elf_file.rb [hunk 8]
```diff
@@ -321,6 +333,54 @@ def save(filename)
...
+      all
+    end
+
+    def append_symbol(*args, **kwargs)
```
Comment (david942j) [inline lib/elftools/elf_file.rb:365]: add descriptions

File: lib/elftools/elf_file.rb [hunk 9]
```diff
@@ -321,6 +333,54 @@ def save(filename)
...
+      symtab.append(*args, **kwargs)
+    end
+
+    def append_section(name, type)
```
Comment (david942j) [inline lib/elftools/elf_file.rb:369]: add descriptions

**Comment by david942j** [comment]
Hi @zygzagZ, thanks for your contributing! The overall idea looks good, but I need you to perform some changes.

First of all, as I can see you are trying to achieve a lot of changes in one PR, I would prefer to separate them to multiple pull requests. Currently this PR is too large that I'm not able to see what's the purpose of each change (because there are too many purposes). A quick glance shows this PR can be separated to at least 5 PRs:
1. The ability of modifying sections
2. Introducing Enum (I didn't think why adding this could make what better, you can add more details when creating this PR)
3. Modification of lazy_array.rb
4. Modification of CHOICE_SIZE_T
5. Bug fix per your PR description

The test failure of posting code coverage is an infra issue, you can ignore them.

Also please add tests when adding new features and keep the test coverage being 100%.

**Review by david942j — CHANGES_REQUESTED** [review]

File: lib/elftools/elf_file.rb (hunk 5 above)
Comment (zygzagZ) [reply lib/elftools/elf_file.rb:345]: could you elaborate?

**Comment by zygzagZ** [comment]
The purpose of Enum was to make it possible to add wrapper methods to Symbol, Section that return header's data in more verbose way.
I wanted to enable usage as such:
```
sym = elf.symtab.symbol_by_name('puts@@GLIBC_2.0')
puts sym.st_type
> "FUNC"
puts "it is a function!" if sym.st_type.func?
sym.st_vis = Visibility.PROTECTED
``
that allows easier debugging without the need of custom functions like `Constants::ET.mapping`.
It is not perfect though, would be nice have const access `Visibility::PROTECTED` or something like `sym.st_vis = :protected`, but did not yet implement this.

**Comment by david942j** [comment]
Hi @zygzagZ , do you have plan to separate your PR to several smaller ones as my previous suggestion?

**Comment by zygzagZ** [comment]
I do, just can't get around to it.
Any tips what would make the least laborious way to achieve it? Is there a tool such that I can decide whether to keep or discard each line of a diff (so as to split it between branches)?

**Comment by david942j** [comment]
Sounds like you need `git add -p` ?

**Comment by zygzagZ** [comment]
superseded by #58 #59 #60 #61 #62 #63 #64.
Hope that nothing falls apart and that I did not miss anything.

**Pull request closed without merging**